### PR TITLE
HDDS-4741. Modularize upgrade test

### PR DESCRIPTION
**Pull request by adoroszlai**
## What changes were proposed in this pull request?

As of now, `upgrade` acceptance test verifies upgrade from Ozone 0.5.0 to "current", ie. the commit being built.  It has an `OZONE_UPGRADE_TO` variable, but its value is not really relevant to which version is run after the "upgrade".

This PR introduces the following changes:

 1. Test upgrade from multiple earlier versions, eg. from 0.5.0 or 1.0.0.
 2. Extract parts of the `upgrade` test script to two libraries (the existing main lib and an upgrade-specific one).
 3. Split `upgrade` test script:
    * `upgrade_to_current`: test upgrade from a release to the current binaries,
    * `upgrade_to_release`: test upgrade from one release to a later one.
 4. Improve the scripts which define version-specific behavior: instead of just declaring variables, define functions to activate/deactivate behavior of a specific version.
 5. Introduce "logical version" to be able to upgrade across multiple versions (eg. from 0.5.0 to 1.1.0).

https://issues.apache.org/jira/browse/HDDS-4741

## How was this patch tested?

Regular CI now runs `upgrade_to_current.sh` from both 0.5.0 and 1.0.0:
https://github.com/adoroszlai/hadoop-ozone/runs/1912441301

Also tested `upgrade_to_release.sh` manually.

### CODE REVIEW

**Comment by avijayanhwx** [comment]
cc @errose28 for review.

**Review by errose28 — COMMENTED** [review]
Thanks for working on this @adoroszlai. Overall looks good, a few minor comments in line. Also it looks like there is an extra robot log and report getting placed in the upgrade directory (outside the results folder). Looks like log and report for 0.5.0 first gets put there, and then log and report for 1.0.0 overwrites it.  Is this supposed to happen?

File: hadoop-ozone/dist/src/main/compose/upgrade/test.sh [hunk 1]
```diff
@@ -14,60 +14,28 @@
...
-if [[ -n "${OZONE_VOLUME_OWNER}" ]]; then
-  current_user=$(whoami)
-  if [[ "${OZONE_VOLUME_OWNER}" != "${current_user}" ]]; then
-    chown -R "${OZONE_VOLUME_OWNER}" "${OZONE_VOLUME}" \
-      || sudo chown -R "${OZONE_VOLUME_OWNER}" "${OZONE_VOLUME}"
+set -x
+SCRIPT_DIR=$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null && pwd )
+ALL_RESULT_DIR="$SCRIPT_DIR/result"
+mkdir -p "$ALL_RESULT_DIR"
+rm "$ALL_RESULT_DIR/*" || true
```
[the remark this answers is not in the document]
Comment (adoroszlai) [reply hadoop-ozone/dist/src/main/compose/upgrade/test.sh:21]: Nice catch.  BTW, this script is copied from `hadoop-ozone/dist/src/main/compose/ozone-mr/test.sh`.

File: hadoop-ozone/dist/src/main/compose/upgrade/test.sh [hunk 2]
```diff
@@ -14,60 +14,28 @@
...
-export OZONE_VOLUME
-
-mkdir -p "${OZONE_VOLUME}"/{dn1,dn2,dn3,om,recon,s3g,scm}
-
-if [[ -n "${OZONE_VOLUME_OWNER}" ]]; then
-  current_user=$(whoami)
-  if [[ "${OZONE_VOLUME_OWNER}" != "${current_user}" ]]; then
-    chown -R "${OZONE_VOLUME_OWNER}" "${OZONE_VOLUME}" \
-      || sudo chown -R "${OZONE_VOLUME_OWNER}" "${OZONE_VOLUME}"
+set -x
```
[the remark this answers is not in the document]
Comment (adoroszlai) [reply hadoop-ozone/dist/src/main/compose/upgrade/test.sh:17]: No, I think it is just leftover debug.

**Comment by adoroszlai** [comment]
> Also it looks like there is an extra robot log and report getting placed in the upgrade directory (outside the results folder). Looks like log and report for 0.5.0 first gets put there, and then log and report for 1.0.0 overwrites it. Is this supposed to happen?

Thanks for pointing this out.  It turns out not specific to the `upgrade` test or this change.  It happens for each "subtest" on `master`, too.  Example (from [here](https://github.com/apache/ozone/runs/1960981044#step:7:107)):

```
Log:     /mnt/ozone/hadoop-ozone/dist/target/ozone-1.1.0-SNAPSHOT/compose/compatibility/result/log.html
Report:  /mnt/ozone/hadoop-ozone/dist/target/ozone-1.1.0-SNAPSHOT/compose/compatibility/result/report.html
Output:  /mnt/ozone/hadoop-ozone/dist/target/ozone-1.1.0-SNAPSHOT/compose/result/compatibility.xml
Log:     /mnt/ozone/hadoop-ozone/dist/target/ozone-1.1.0-SNAPSHOT/compose/log.html
Report:  /mnt/ozone/hadoop-ozone/dist/target/ozone-1.1.0-SNAPSHOT/compose/report.html
```

(The second log and report are the ones created in the same way as you described.)

Since the fix is [pretty simple](https://github.com/apache/ozone/pull/1928/commits/749d47121886d282e55a5e5b1f1de47031bfb330), I incuded it here, instead of creating a separate issue.

**Comment by errose28** [comment]
Thanks for working on this @adoroszlai. Changes LGTM +1.

**Comment by adoroszlai** [comment]
@avijayanhwx @elek @fapifta can you please review?  @errose28 is OK with it and would like to work on top of it in the upgrade branch.

**Review by elek — APPROVED** [review]
+1 LGTM (except one typo). 

Thanks the patch @adoroszlai

File: hadoop-ozone/dist/src/main/compose/versions/README.md [hunk 3]
```diff
@@ -13,3 +13,11 @@
 -->
 
 The scripts in this directory define version-specific behavior required for [`testlib.sh`](../../testlib.sh).  For example the `ozone admin` command was renamed from `ozone scmcli` in 1.0.0.
```
[the remark this answers is not in the document]
Comment (adoroszlai) [reply hadoop-ozone/dist/src/main/compose/versions/README.md]: Thanks!

**Review by avijayanhwx — APPROVED** [review]
Thanks for working on this @adoroszlai. I have gone through the changes and they look good to me.

**Comment by avijayanhwx** [comment]
Thanks for the patch @adoroszlai, and the reviews @errose28 & @elek. I am merging this.

**Pull request merged**